modules/basics/src/main/java/com/opengamma/basics/currency/FxMatrix.java

```diff
@@ -0,0 +1,557 @@
+   * should have at least one currency in common with this one.
+   * The additional currencies from the other matrix are added one by one
+   * and the exchange rate data created is coherent with some data in this
+   * matrix. If the data in the two matrices are not coherent then there is
+   * no guarantee which data will be used and the final result may be incoherent.
+   *
+   * @param other  the matrix to be merged into this one
+   * @return a new matrix containing the rates from this matrix
+   *   plus any rates for additional currencies from the other matrix
+   */
+  public FxMatrix merge(FxMatrix other) {
+    return toBuilder().merge(other.toBuilder()).build();
+  }
+
+  /**
+   * Returns an immutable set containing the currencies held within this matrix.
+   *
+   * @return the currencies in this matrix
+   */
+  public ImmutableSet<Currency> getCurrencies() {
+    return currencies.keySet();
+  }
+
+  @Override
+  public String toString() {
+    return getCurrencies() + " - " + Stream.of(rates).map(Arrays::toString).collect(Collectors.joining());
+  }
+
+  @Override
+  public int hashCode() {
+    return 31 * currencies.hashCode() + Arrays.deepHashCode(rates);
+  }
+
+  @Override
+  public boolean equals(final Object obj) {
+    if (this == obj) {
+      return true;
+    }
+    if (obj == null || getClass() != obj.getClass()) {
+      return false;
+    }
+    FxMatrix other = (FxMatrix) obj;
+    return currencies.equals(other.currencies) && Arrays.deepEquals(rates, other.rates);
+  }
+
+  /**
+   * Creates a {@link Collector} that allows a Map of currency pair -> rates
+   * to be streamed and collected into a new {@code FxMatrix}.
+   *
+   * @return a collector for creating an {@code FxMatrix} from a stream
+   */
+  public static Collector<? super Pair<CurrencyPair, Double>, Builder, FxMatrix> pairsToFxMatrix() {
+    return collector((builder, pair) ->
+        builder.addRate(pair.getFirst(), pair.getSecond()));
+  }
+
+  /**
+   * Creates a {@link Collector} that allows a collection of pairs each containing
+   * a currency pair and a rate to be streamed and collected into a new {@code FxMatrix}.
+   *
+   * @return a collector for creating an {@code FxMatrix} from a stream
+   */
+  public static Collector<? super Map.Entry<CurrencyPair, Double>, Builder, FxMatrix> entriesToFxMatrix() {
+    return collector((builder, entry) ->
+        builder.addRate(entry.getKey(), entry.getValue()));
+  }
+
+  private static <T> Collector<T, Builder, FxMatrix> collector(BiConsumer<Builder, T> accumulator) {
+    return Collector.of(
+        FxMatrix::builder,
+        accumulator,
+        Builder::merge,
+        Builder::build);
+  }
+
+  /**
+   * Builder class for FxMatrix. Can be created either by the static
+   * {@link FxMatrix#builder()} or from an existing {@code FxMatrix}
+   * instance by calling {@link FxMatrix#toBuilder()}.
+   */
+  public static final class Builder {
```

Comment:
This will have to become a top-level class when ``FxMatrix`` becomes a bean and has a generated ``Builder`` class.